governance/program/src/state/proposal.rs

```diff
@@ -229,25 +280,84 @@ impl Proposal {
         Ok(())
     }
 
+    /// Returns final proposal state after vote ends
     fn get_final_vote_state(
         &mut self,
         max_vote_weight: u64,
         config: &GovernanceConfig,
-    ) -> ProposalState {
-        let yes_vote_threshold_count =
-            get_yes_vote_threshold_count(&config.vote_threshold_percentage, max_vote_weight)
+    ) -> Result<ProposalState, ProgramError> {
+        // Get the min vote weight required for options to pass
+        let min_vote_threshold_weight =
+            get_min_vote_threshold_weight(&config.vote_threshold_percentage, max_vote_weight)
                 .unwrap();
 
-        // Yes vote must be equal or above the required yes_vote_threshold_percentage and higher than No vote
-        // The same number of Yes and No votes is a tie and resolved as Defeated
-        // In other words  +1 vote as a tie breaker is required to Succeed
-        if self.yes_votes_count >= yes_vote_threshold_count
-            && self.yes_votes_count > self.no_votes_count
-        {
-            ProposalState::Succeeded
-        } else {
+        // If the proposal has a reject option then any other option must beat it regardless of the configured min_vote_threshold_weight
+        let deny_vote_weight = self.deny_vote_weight.unwrap_or(0);
+
+        let mut best_succeeded_option_weight = 0;
+        let mut best_succeeded_option_count = 0;
+
+        for option in self.options.iter_mut() {
+            // Any positive vote (Yes) must be equal or above the required min_vote_threshold_weight and higher than the reject option vote (No)
+            // The same number of positive (Yes) and rejecting (No) votes is a tie and resolved as Defeated
+            // In other words  +1 vote as a tie breaker is required to succeed for the positive option vote
+            if option.vote_weight >= min_vote_threshold_weight
+                && option.vote_weight > deny_vote_weight
```

Comment:
I had the same idea (and implemented it like that) at the beginning but then realised it needed to be separate for several reasons
- There would have to be either a flag on an option to indicate `Deny` choice or some convention ex.  the last option is `Deny`. In both cases some special treatment and conditional coding and extra validation was required. 
- It opened a potential vector of attack. A malicious user could attach an instruction to the `Deny` option and hence execute code on both Yes/No outcomes of the vote. It could be guarded through assertions but with a separate option that it's not even possible. 
- Some small savings in rent because `Deny` doesn't need all the data (label, result, instructions) as positive options
- Overall it felt more dirty and not as straightforward as the explicit `Deny` option

Reply:
That makes sense, thanks for the explanation